File: listas/src/main/java/com/victorlh/android/framework/listas/adapter/ListaObservableAdapter.java
package com.victorlh.android.framework.listas.adapter;

import androidx.annotation.NonNull;

import com.victorlh.android.framework.listas.ItemLista;

import java.util.ArrayList;
import java.util.List;

class ListaObservableAdapter<T extends ItemLista> extends ListaObservable<T> {

	private List<ListaObservable<T>> lista;

	ListaObservableAdapter() {
		this.lista = new ArrayList<>();
	}

	void registrarObservable(ListaObservable<T> observable) {
		lista.add(observable);
	}

	void removerObservable(ListaObservable<T> observable) {
		lista.remove(observable);
	}

	protected void setItem(T item) {
		for (int i = lista.size() - 1; i >= 0; i--) {
			ListaObservable<T> observable = lista.get(i);
			observable.setItem(item);
		}
	}

	protected void deleteItem(T item) {
		for (int i = lista.size() - 1; i >= 0; i--) {
			ListaObservable<T> observable = lista.get(i);
			observable.deleteItem(item);
		}
	}

	protected void recargarLista() {
		for (int i = lista.size() - 1; i >= 0; i--) {
			ListaObservable<T> observable = lista.get(i);
			observable.recargarLista();
		}
	}

	protected boolean onClickElemento(AbstractViewHolder<T> viewHolder) {
		boolean interceptado = false;
		for (int i = lista.size() - 1; i >= 0; i--) {
			ListaObservable<T> observable = lista.get(i);
			boolean aux = observable.onClickElemento(viewHolder);
			if (aux) {
				interceptado = true;
			}
		}
		return interceptado;
	}

	protected boolean onLongClickElemento(AbstractViewHolder<T> viewHolder) {
		boolean interceptado = false;
		for (int i = lista.size() - 1; i >= 0; i--) {
			ListaObservable<T> observable = lista.get(i);
			boolean aux = observable.onLongClickElemento(viewHolder);
			if (aux) {
				interceptado = true;
			}
		}
		return interceptado;
	}

	protected void onBindViewHolder(@NonNull final AbstractViewHolder<T> holder, int position) {
		for (int i = lista.size() - 1; i >= 0; i--) {
			ListaObservable<T> observable = lista.get(i);
			observable.onBindViewHolder(holder, position);
		}
	}
}

File: listas/src/main/java/com/victorlh/android/framework/listas/adapter/ListaObservableAdapter.kt
package com.victorlh.android.framework.listas.adapter

import java.util.*

internal class ListaObservableAdapter<T> : ListaObservable<T>() {

    private val lista: MutableList<ListaObservable<T>>

    fun registrarObservable(observable: ListaObservable<T>) {
        lista.add(observable)
    }

    fun removerObservable(observable: ListaObservable<T>?) {
        lista.remove(observable)
    }

    override fun setItem(item: T) {
        for (i in lista.indices.reversed()) {
            val observable = lista[i]
            observable.setItem(item)
        }
    }

    override fun deleteItem(item: T) {
        for (i in lista.indices.reversed()) {
            val observable = lista[i]
            observable.deleteItem(item)
        }
    }

    override fun recargarLista() {
        for (i in lista.indices.reversed()) {
            val observable = lista[i]
            observable.recargarLista()
        }
    }

    override fun onClickElemento(viewHolder: AbstractViewHolder<T>): Boolean {
        var interceptado = false
        for (i in lista.indices.reversed()) {
            val observable = lista[i]
            val aux = observable.onClickElemento(viewHolder)
            if (aux) {
                interceptado = true
            }
        }
        return interceptado
    }

    override fun onLongClickElemento(viewHolder: AbstractViewHolder<T>): Boolean {
        var interceptado = false
        for (i in lista.indices.reversed()) {
            val observable = lista[i]
            val aux = observable.onLongClickElemento(viewHolder)
            if (aux) {
                interceptado = true
            }
        }
        return interceptado
    }

    override fun onBindViewHolder(holder: AbstractViewHolder<T>, position: Int) {
        for (i in lista.indices.reversed()) {
            val observable = lista[i]
            observable.onBindViewHolder(holder, position)
        }
    }

    init {
        lista = ArrayList()
    }
}